GDJS/Runtime/websocket-debugger-client/hot-reloader.js

```diff
@@ -0,0 +1,1313 @@
+      );
+    }
+  });
+  oldVariablesData.forEach((oldVariableData) => {
+    const newVariableData = newVariablesData.filter(
+      (variable) => variable.name === oldVariableData.name
+    )[0];
+
+    if (!newVariableData) {
+      // Variable was removed
+      variablesContainer.remove(oldVariableData.name);
+    }
+  });
+};
+
+/**
+ * @param {VariableData[]?} oldChildren
+ * @param {VariableData[]} newChildren
+ * @param {gdjs.Variable} variable
+ */
+gdjs.HotReloader.prototype._hotReloadStructureVariable = function (
+  oldChildren,
+  newChildren,
+  variable
+) {
+  if (oldChildren) {
+    oldChildren.forEach((oldChildVariableData) => {
+      const newChildVariableData = newChildren.filter(
+        (childVariableData) =>
+          childVariableData.name === oldChildVariableData.name
+      )[0];
+
+      if (!newChildVariableData) {
+        // Child variable was removed.
+        variable.removeChild(oldChildVariableData.name);
+      } else if (
+        !newChildVariableData.children &&
+        (oldChildVariableData.value !== newChildVariableData.value ||
+          !oldChildVariableData.children)
+      ) {
+        // The child variable value was changed or was converted from
+        // structure to a variable with value.
+        variable.addChild(
+          newChildVariableData.name,
+          new gdjs.Variable(newChildVariableData)
+        );
+      } else if (newChildVariableData.children) {
+        // The child variable is a structure.
+        this._hotReloadStructureVariable(
+          oldChildVariableData.children,
+          newChildVariableData.children,
+          variable.getChild(newChildVariableData.name)
+        );
+      }
+    });
+    newChildren.forEach((newChildVariableData) => {
+      const oldChildVariableData = oldChildren.filter(
+        (childVariableData) =>
+          childVariableData.name === newChildVariableData.name
+      )[0];
+
+      if (!oldChildVariableData) {
+        // Child variable was added
+        variable.addChild(
+          newChildVariableData.name,
+          new gdjs.Variable(newChildVariableData)
+        );
+      }
+    });
+  } else {
+    // Variable was converted from a value to a structure:
+    newChildren.forEach((newChildVariableData) => {
+      variable.addChild(
+        newChildVariableData.name,
+        new gdjs.Variable(newChildVariableData)
+      );
+    });
+  }
+};
+
+/**
+ * @param {LayoutData} oldLayoutData
+ * @param {LayoutData} newLayoutData
+ * @param {ChangedRuntimeBehavior[]} changedRuntimeBehaviors
+ * @param {gdjs.RuntimeScene} runtimeScene
+ */
+gdjs.HotReloader.prototype._hotReloadRuntimeScene = function (
+  oldLayoutData,
+  newLayoutData,
+  changedRuntimeBehaviors,
+  runtimeScene
+) {
+  runtimeScene.setBackgroundColor(
+    newLayoutData.r,
+    newLayoutData.v,
+    newLayoutData.b
+  );
+  if (oldLayoutData.title !== newLayoutData.title) {
+    runtimeScene.getGame().getRenderer().setWindowTitle(newLayoutData.title);
```

Comment:
@Bouh Notice how the hot-reloader is always using public methods to do changes to the game (you could do all that is done by the hot reloader using JavaScript code events). 
There is nothing "implicit" and no hidden magic in the sense that the hot-reloader is like an automatic assistant that is calling various functions of the game/scene/objects to reflect the difference it sees between before and after.
A very useful assistant :)